Add sampling support

jjatria opened the pull request:
The main point of this change is to add support for using a sampling rate with the current functions. An `IncSampled` was added in 98dec13 but no equivalent functions for `timer.Send`, `Time`, or `Gauge` were added.

A problem with adding these as new functions is that `IncSampled` placed the sampling rate argument at the end of the signature, but some of our existing functions (eg. `timer.Send`) take a variadic list of timer names, so that would not work.

This PR solves this by adding a `statsd.Options` struct to hold options such as the desired sample rate, and a new set of `*WithOptions` functions to go with it that take a `*statsd.Options` as their first argument (which means that they should be callable with our defaults by passing `nil`). A `statsd.DefaultOptions()` function is also exported, to provide a copy of those defaults.

This also includes an expanded test suite, done in the least efficient way possible for now. This is ripe for later improvements.

The changes in this PR should be backwards compatible, and the `IncSampled` function remains unchanged, but with a deprecation notice.

jjatria committed:
Simplify increment code

jjatria committed:
Move sampling rate to start of signature

jjatria committed:
Implement consistent sampling methods

jjatria committed:
Expand test coverage

jjatria committed:
Sample data before sending over the wire

jjatria committed:
Move sampling to Options struct

jjatria committed:
Restore IncSampled and mark as deprecated

This is here solely to maintain backwards compatibility

jjatria committed:
Document remaining exported symbols

colinnewell commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
+	// Rate specifies the sampling rate to use.
+	Rate float64
+	// AlwaysSend specifies whether the packet should be sent
+	// to the server even if the specified sampling rate is too low.
```
too low?

colinnewell commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
 }
 
+// Send takes a list of remote timer names, and submits the time that
+// has ellapsed since the creation of the timer to each in turn.
```
elapsed

skyerus replied:
typo `elapsed`

colinnewell commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
+}
+
+// SendWithOptions works like Send but sends the timing information
+// using the provided options.
```
could we add examples? 
https://blog.golang.org/examples

jjatria replied:
I've added a bunch of examples, and tested those with a local version of `godoc`. In order for the `timer.Send` functions to render, though, I had to export the struct, which I did as `Timing`, to avoid the conflict with the `Timer` function.

colinnewell commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
+// Deprecated: Use IncWithOptions and pass the sampling rate
+// using the Options struct.
+func IncSampled(name string, rate float64) {
+	IncWithOptions(&Options{Rate: rate}, name)
```
should this not always send otherwise you're changing the behaviour?

jjatria replied:
True. I'll add that

jjatria committed:
Update README and documentation

skyerus commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
+// using the provided options.
+func (t *timer) SendWithOptions(
+	options *Options,
+	names ...interface{},
```
why are `names` of type `interface` when they're expected to be cast to strings?

jjatria replied:
Mainly for consistency with `Send(names ...interface{})`. I agree with you that this should probably have always been `...string`, but I feel like that ship has sailed.

skyerus commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
-	// Send a host suffixed stat too.
-	if AlsoAppendHost {
-		conn.Write([]byte(name + "." + host + value))
+	if !options.AlwaysSend && rand.Float64() >= options.Rate {
```
Perhaps `rand` should be seeded in `init()`

skyerus commented on statsd.go:
```diff
@@ -25,110 +26,159 @@ func init() {
+}
+
+// DefaultOptions returns the default options
+func DefaultOptions() *Options {
```
Perhaps this could be an exported var instead, allowing the default options to be overwritten

jjatria replied:
Ah, yeah, I guess that works. I moved it to a function because I was trying to set it as a `const`, but making it settable does make sense.

jjatria committed:
Export Timing struct

This will allow the documentation for the related methods
to be rendered in GoDoc.

The struct was named "Timing" to not collide with the "Timer"
function.

jjatria committed:
Add examples

jjatria commented:
I don't think you can have examples for individual methods for a struct. At least I didn't figure out what magic name to give to the `Example...` function for it to render properly. It _does_ seem like you can have examples for the exported structs, so I added a `Timing` example that covers most bases.

I tried several ways to split them apart, and none of them worked. If there are ways to split them, I'll be happy to use them.
